### PR TITLE
Add migrate support to `pallet-cosmwasm` and implement migrate entrypoint on contracts

### PR DESCRIPTION
**Pull request by aeryz**
## Description
* Add `migrate` extrinsic for contract migration.
* Add `migrate` entrypoint to all of the cosmwasm contracts as well as CONTRACT_VERSION and NAMES to make sure that migration is securely done.
* Add `clear_admin` extrinsic to cover `ClearAdmin` functionality in cosmwasm.
* Add `update_admin` extrinsic to cover `UpdateAdmin` functionality in cosmwasm.
* Add `verify`s to extrinsic's benchmarks to test them during benchmarking.
* Use `CodeIdentifier` instead of code id to make users able to use code hash if they want to for example batch instantiate + execute or migrate calls. This way they would be able to identify the code in-front.

## Checklist

- [X] I have updated the cargo docs to reflect changes made by this PR _(if applicable)_
- [X] I have updated the `book/` to reflect changes made by this PR _(if applicable)_

### CODE REVIEW

File: code/parachain/frame/cosmwasm/src/lib.rs [hunk 1]
```diff
@@ -507,6 +555,58 @@ pub mod pallet {
...
+				Ok(())
+			})?;
+
+			Self::deposit_event(Event::<T>::Migrated {
+				contract: contract.clone(),
+				to: new_code_id,
+			});
+
+			Ok(contract_info)
+		}
```
Comment (KaiserKarel) [inline code/parachain/frame/cosmwasm/src/lib.rs:608]: Nit: this pattern allows for incorrectly migrating a contract. Instead we could use a Migrator struct, using Typestate to ensure that `migrate_phase1` is called before a contract is migrated.

Comment (KaiserKarel) [inline code/parachain/frame/cosmwasm/src/lib.rs:608]: Nit: this pattern allows for incorrectly migrating a contract. Instead we could use a Migrator struct, using Typestate to ensure that `migrate_phase1` is called before a contract is migrated.

Comment (KaiserKarel) [reply code/parachain/frame/cosmwasm/src/lib.rs:608]: In general, let's avoid patterns where foo_1 needs to be called before foo, and encode that logic in the type system.

**Review by KaiserKarel — CHANGES_REQUESTED** [review]
Still needs benchmarks

File: code/xcvm/cosmwasm/contracts/asset-registry/src/contract.rs [hunk 2]
```diff
@@ -32,6 +38,12 @@ pub fn execute(
 	}
 }
 
+#[cfg_attr(not(feature = "library"), entry_point)]
+pub fn migrate(deps: DepsMut, _env: Env, _msg: MigrateMsg) -> Result<Response, ContractError> {
```
Comment (dzmitry-lahoda) [inline code/xcvm/cosmwasm/contracts/asset-registry/src/contract.rs:42]: please add /// doc to it.

**Review by dzmitry-lahoda — APPROVED** [review]

**Review by KaiserKarel — APPROVED** [review]

File: code/parachain/frame/cosmwasm/src/lib.rs [hunk 3]
```diff
@@ -674,21 +749,70 @@ pub mod pallet {
...
+			if info.code_id != code_id {
+				// Increase the refcount of `new_code_id`.
+				CodeIdToInfo::<T>::try_mutate_exists(code_id, |entry| -> Result<(), Error<T>> {
+					let code_info = entry.as_mut().ok_or(Error::<T>::CodeNotFound)?;
+					code_info.refcount =
+						code_info.refcount.checked_add(1).ok_or(Error::<T>::RefcountOverflow)?;
+					Ok(())
+				})?;
+
+				// Modify the exising `code_id`'s states and unreserve the bonded funds.
```
Comment (Martin-Kiss) [inline code/parachain/frame/cosmwasm/src/lib.rs:776]:
```suggestion
				// Modify the existing `code_id`'s states and unreserve the bonded funds.
```
Nit spelling; failing spellcheck CI

**Pull request merged**